### DISCUSSION TITLE
How to get an AWS S3 object from an encrypted bucket

### DISCUSSION THREAD

**Discussion by rocketnova** [lowdefy/lowdefy#1563] [Questions and Help] [answered]
I'm trying to use the [AWS S3](https://docs.lowdefy.com/AWSS3) connection in lowdefy version 3.23.3. I have an encrypted S3 bucket. I can generate a presigned link, but I get the following error when accessing the link:

```xml
<Error>
  <Code>InvalidArgument</Code>
  <Message>
    Requests specifying Server Side Encryption with AWS KMS managed keys require AWS Signature Version 4.
  </Message>
  <ArgumentName>Authorization</ArgumentName>
  <ArgumentValue>null</ArgumentValue>
  <RequestId>...</RequestId>
  <HostId>...</HostId>
</Error>
```

To generate the correct presigned url, in [AwsS3PresignedGetObject.js](https://github.com/lowdefy/lowdefy/blob/main/packages/plugins/plugins/plugin-aws/src/connections/AwsS3Bucket/AwsS3PresignedGetObject/AwsS3PresignedGetObject.js), we would need to pass in an explicit `signatureVersion`:

```
const s3 = new AWS.S3({ accessKeyId, secretAccessKey, region, bucket, signatureVersion: "v4" });
```

What's the best way for me to accomplish this in lowdefy v3? Do I need to create a custom ... connection?

**Comment by rocketnova** [comment]
I suspect the answer is that this would be a lot easier in v4 and would be solved by using the plugin system to add a [custom connection](https://docs-git-docs-plugins-lowdefy.vercel.app/plugins-connections). Is that true?

I actually want to dynamically generate a number of presigned s3 links on page load based on the results from a db query. But I haven't been able to identify a way to loop through an array to dynamically define requests in lowdefy on page load. Is that possible?

Thank you for your help!

**Reply by SamTolmay** [reply, marked as answer]
Hi @rocketnova 

Yes a custom connection would be a good fit here, especially if you would like to loop through a list of objects. We are looking at ways to do loops in Lowdefy config, but we have found custom requests to be a very easy way to execute some specific server-side code - The custom request then functions like an API endpoint function.

We can add the `signatureVersion` as a property, but this will also be on V4, we are not adding new features on V3 and will be releasing V4 soon - it is ready for use, we are just finalising documentation.

**Reply by rocketnova** [reply]
@SamTolmay Thank you very much for your response. I'm going to solve this problem a different way for now, but I hope to have time to revisit this when V4 is released and migrate my tool to take advantage of the cool new features.
